homeassistant/components/variables/__init__.py

```diff
@@ -0,0 +1,226 @@
+
+DOMAIN = "variables"
+
+STORAGE_KEY = DOMAIN
+STORAGE_VERSION = 1
+STORAGE_SAVE_DELAY = 10
+
+DATA_STORAGE = "storage"
+
+TOPIC_VARIABLE_SET = "variables_set_{0}"
+
+UNIQUE_ID_TEMPLATE = "variables_{0}"
+
+CONF_ATTRS = "attributes"
+CONF_KEY = "key"
+CONF_VALUE = "value"
+
+REMOVE_VARIABLE_SCHEMA = vol.Schema({vol.Required(CONF_KEY): cv.string})
+SET_VARIABLE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_KEY): cv.string,
+        vol.Required(CONF_VALUE): cv.string,
+        vol.Optional(CONF_ATTRS): dict,
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Variables component."""
+    variable_manager = VariableManager(hass)
+    await variable_manager.async_init()
+
+    entity_component = EntityComponent(_LOGGER, DOMAIN, hass)
+
+    async def async_remove_variable(call: ServiceCall):
+        """Remove a variable."""
+        key = call.data[CONF_KEY]
+
+        variable_manager.async_remove(key)
+
+        ent_reg = await entity_registry.async_get_registry(hass)
+        ent_to_remove = ent_reg.async_get_entity_id(
+            DOMAIN, DOMAIN, UNIQUE_ID_TEMPLATE.format(key)
+        )
+        if ent_to_remove is not None:
+            ent_reg.async_remove(ent_to_remove)
+
+    async def async_set_variable(call: ServiceCall):
+        """Set a variable."""
+        key = call.data[CONF_KEY]
+        value = call.data[CONF_VALUE]
+        attributes = call.data.get(CONF_ATTRS)
+
+        variable = await variable_manager.async_set(key, value, attributes=attributes)
+
+        ent_reg = await entity_registry.async_get_registry(hass)
+        existing_ent = ent_reg.async_get_entity_id(
+            DOMAIN, DOMAIN, UNIQUE_ID_TEMPLATE.format(key)
+        )
+        if existing_ent:
+            async_dispatcher_send(hass, TOPIC_VARIABLE_SET.format(key))
+        else:
+            await entity_component.async_add_entities([VariableEntity(variable)])
+
+    for service, method, schema in [
+        ("remove", async_remove_variable, REMOVE_VARIABLE_SCHEMA),
+        ("set", async_set_variable, SET_VARIABLE_SCHEMA),
+    ]:
+        hass.services.async_register(DOMAIN, service, method, schema=schema)
+
+    return True
+
+
+class VariableStore(Store):
+    """Define variable storage."""
+
+    async def _async_migrate_func(self, old_version, old_data):
+        """Migrate to the new version.
+
+        Migrate storage to use format of collection helper.
+        """
+        return old_data
+
+
+@dataclass
+class Variable:
+    """Define a variable."""
+
+    key: str
+    value: str
+    attributes: Optional[dict]
+
+
+class VariableManager:
+    """Define an object to manage variables."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize."""
+        self._storage: VariableStore = VariableStore(hass, STORAGE_VERSION, STORAGE_KEY)
+        self._variables: Dict[str, Variable] = {}
+
```

Comment:
This component is written with a lot of async functions, however, there is really nothing to await in the end. 

Not sure whether it's more efficient this way. Anyone?

Reply:
You're right about `async_set` – fixed that in 90700aa. Any others?

Reply:
BTW I pinged you on Discord. Is this a "Draft" because you want to work on it before people give you feedback **or** is it a "Draft" because you just don't consider it to be 100% done?

Reply:
No need for draft status since this works.